## NumPy Mathematical Operations, Equations & Sequences
### Question 1: Business Analytics — Sales Performance Calculator

In [27]:
# The management team wants to understand the week's performance.
# The business recorded the following daily sales for one week:

import numpy as np

daily_sales = np.array([125000, 150000, 175000, 140000, 190000, 210000, 160000])
print()
print("Daily Sales:", daily_sales)
print(type(daily_sales))
print(daily_sales.shape)

# To calculate total sales 
total_sales = np.sum(daily_sales)
print("Total sales:", total_sales)

# To the Calculate Average Daily Sales
average_sales = np.mean(daily_sales)
print()
print("Average Sales:", average_sales)

# Apply a Sales Adjustment
adjusted_sales = daily_sales * 1.10
print()
print("Adjusted Sales:", adjusted_sales * 10)

# Task 5  Compare Original and Adjusted Sales
daily_sales = np.array([125000, 150000, 175000, 140000, 190000, 210000, 160000]) # original sales
adjusted_sales = daily_sales * 1.10  # Adjusted sales
print()
print("Original Sales:", daily_sales)
print("Adjusted Sales:", adjusted_sales)





Daily Sales: [125000 150000 175000 140000 190000 210000 160000]
<class 'numpy.ndarray'>
(7,)
Total sales: 1150000

Average Sales: 164285.7142857143

Adjusted Sales: [1375000. 1650000. 1925000. 1540000. 2090000. 2310000. 1760000.]

Original Sales: [125000 150000 175000 140000 190000 210000 160000]
Adjusted Sales: [137500. 165000. 192500. 154000. 209000. 231000. 176000.]


### Looking at the above result the question is to know What does the average daily sales figure tell the business that the total sales figure does not?

#### Total sales: 1150000
#### Average Sales: 164285.71

### Explanation

#### The total sales figure shows the business that it generated ₦1,150,000 in sales over the 7-day period.

#### The average daily sales figure shows the business that it generated approximately ₦164,285.71 per day on average.

#### The average is useful because it helps the business understand its typical daily sales performance, while the total only shows the overall sales for the entire period.

## Exercise 2: Student Performance Analysis

### The program coordinator wants to understand the overall performance of the students and how individual scores differ from the group average.


In [2]:
# Task 1 — Create and Inspect the Data
import numpy as np

scores = np.array([62, 75, 81, 69, 88, 94, 73, 85, 77, 91]) 
print("Score:", scores)
print(type(scores))
print(scores.shape)
print()

#Task 2 — Calculate the Mean Score
mean_score = np.mean(scores)
print("Mean Score:", mean_score)
print()

# Task 3 — Calculate Each Student's Deviation
deviations = scores - mean_score
print("Deviations:", deviations)
print()

#Task 5 — Square the Deviations
square_deviations = deviations ** 2
print("Square Deviation:", square_deviations)
print()
# # Squaring the deviations changes negative values into positive values
# because each deviation is multiplied by itself.

# Task 6 — Calculate the Sum of Squared Deviations
sum_squared_deviations = np.sum(square_deviations)
print("Sum squared deviations:", sum_squared_deviations)
print()


Score: [62 75 81 69 88 94 73 85 77 91]
<class 'numpy.ndarray'>
(10,)

Mean Score: 79.5

Deviations: [-17.5  -4.5   1.5 -10.5   8.5  14.5  -6.5   5.5  -2.5  11.5]

Square Deviation: [306.25  20.25   2.25 110.25  72.25 210.25  42.25  30.25   6.25 132.25]

Sum squared deviations: 932.5



## Task 7 — Connect the Steps to Standard Deviation

### The stages involved in calculating standard deviation are:

####  Calculate the mean
#### Calculate the deviations from the mean
#### Square the deviations
#### Sum the squared deviations
#### Divide by the appropriate number of values
#### Take the square root

#### Standard deviation helps us understand how spread out the students' scores are from the mean score.
#### A smaller standard deviation means the scores are closer to the mean, 
#### while a larger standard deviation means the scores are more spread out.

## HANDS-ON EXERCISE 3
### Engineering & Scientific Computing — Trigonometric Series and Convergence

In [31]:
# Task 1 — Define the Angle
theta = 30   #This represents an angle in degrees.
theta_radian = np.radians(theta)

print("Theta Radian:", theta_radian)
print()

# Task 2: Task 2 — Create the Sequence of Terms
k = np.arange(1, 100, 1)
print(k)
print()

#Task 3 — Calculate the Trigonometric Component
sine_value = np.sin(theta_radian)
print("sine Value:", sine_value)
print()

# Task 4 — Build the Series Terms
series_terms = np.sin(theta_radian) / k
print(series_terms)
print()

# Task 5 — Calculate the Sum
sum_total = np.sum(series_terms)
print("Total Sum:", sum_total)
print()

# Task 6 — Increase the Number of Terms
k = np.arange(1, 100, 1)
series_terms = np.sin(theta_radian) / k
sum_total = np.sum(series_terms)

print("sum 99 Total:", sum_total)

k = np.arange(1, 1000, 1)
series_terms = np.sin(theta_radian) / k
sum_total = np.sum(series_terms)

print("sum 999 Total:", sum_total)


k = np.arange(1, 10000, 1)
series_terms = np.sin(theta_radian) / k
sum_total = np.sum(series_terms)

print("sum 9,999 Total:", sum_total)



Theta Radian: 0.5235987755982988

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72
 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96
 97 98 99]

sine Value: 0.49999999999999994

[0.5        0.25       0.16666667 0.125      0.1        0.08333333
 0.07142857 0.0625     0.05555556 0.05       0.04545455 0.04166667
 0.03846154 0.03571429 0.03333333 0.03125    0.02941176 0.02777778
 0.02631579 0.025      0.02380952 0.02272727 0.02173913 0.02083333
 0.02       0.01923077 0.01851852 0.01785714 0.01724138 0.01666667
 0.01612903 0.015625   0.01515152 0.01470588 0.01428571 0.01388889
 0.01351351 0.01315789 0.01282051 0.0125     0.01219512 0.01190476
 0.01162791 0.01136364 0.01111111 0.01086957 0.0106383  0.01041667
 0.01020408 0.01       0.00980392 0.00961538 0.00943396 0.00925926
 0.00909091 0.00892857 0.0

### Task 7 — Investigate Convergence

### | Number of Terms | Series Result |
### | 99              | 2.5887 |
### | 999             | 3.7422 |
### | 9,999           | 4.8938|

#### Observation

#### As the number of terms increases, the series result also increases.
#### However, the increase becomes smaller as more terms are added.
#### This shows that the series is gradually approaching a stable value.

### Task 8 — Explain Convergence

###Convergence means that a result gets closer and closer to a particular value as more terms are added to a series.

#### What happened when I used fewer terms? 
##### When I used fewer terms, the series result was a smaller approximation because only a limited number of terms were included.

#### What happened when I increased the number of terms?
##### When I increased the number of terms, the series result increased and became a better approximation of the series.

#### Why does increasing the number of terms change the approximation?* 
##### Increasing the number of terms means that more values are added to the series. These additional terms contribute to the total, so the result changes.

#### What does it mean when the result gets closer to a limiting value?**  
##### It means that the series is converging. As more terms are added, the result gets closer to a particular value, called the limiting value.

# Bonus Challenge
### Monthly Business Expense Analysis Using NumPy

## Scenario
### A small business wants to analyze its monthly expenses using NumPy.
### The business records expenses for January to June and wants to understand
### its total expenses, average monthly expense, and how expenses change from month to month.

## Objective
### The objective of this experiment is to use NumPy to calculate and analyze
### monthly business expenses and understand how numerical operations can helpwith simple financial analysis.

In [4]:
import numpy as np

# Monthly business expenses from January to June
expenses = np.array([120000, 135000, 128000, 150000, 142000, 160000])

print("Monthly Expenses:", expenses)
print()

# To calculate total expenses
total_expenses = np.sum(expenses)

print("Total Expenses:", total_expenses)
print()

average_expense = np.mean(expenses)

print("Average Monthly Expense:", average_expense)   

Monthly Expenses: [120000 135000 128000 150000 142000 160000]

Total Expenses: 835000

Average Monthly Expense: 139166.66666666666


## Interpretation

#### The analysis shows the total amount spent by the business over the six-monthperiod and the average monthly expense.
#### The monthly expense changes show how spending increased or decreased from onemonth to the next. This can help the business identify months with higher expenses and monitor changes in its spending pattern.

## Conclusion

#### This experiment helped me apply NumPy to a simple real-world financial scenario. I used NumPy arrays, sum, mean, and subtraction to analyze monthly business expenses.
#### The exercise showed me how NumPy can make numerical calculations easier and more efficient when working with data.

